Add sample for ESPRESSIF ESP32 Azure IoT Kit (#97)
Azure-Samples/iot-middleware-freertos-samples: demos/projects/ESPRESSIF/aziotkit/README.md

# Connect an ESPRESSIF ESP32 using Azure IoT Middleware for FreeRTOS

## What you need

* [ESPRESSIF ESP32 Board](https://www.espressif.com/en/products/devkits)

> * USB 2.0 A male to Micro USB male cable

Reviewer: ```suggestion
* Micro USB male cable
```

Reviewer: you can't assume USB-A, I can use USB-C to micro UBS on my surface :)

* WiFi Connection

* [ESP-IDF](https://idf.espressif.com/) (Version 4.3 for Microsoft Windows or 4.4 for Linux)

* To run this sample you can use a device previously created in your IoT Hub or have the Azure IoT Middleware for FreeRTOS provision your device automatically using DPS.

  IoT Hub | DPS
  ---------|----------
  Have an [Azure IoT Hub](https://docs.microsoft.com/azure/iot-hub/iot-hub-create-through-portal) created | Have an instance of [IoT Hub Device Provisioning Service](https://docs.microsoft.com/en-us/azure/iot-dps/quick-setup-auto-provision#create-a-new-iot-hub-device-provisioning-service)
  Have a [logical device](https://docs.microsoft.com/azure/iot-hub/iot-hub-create-through-portal#register-a-new-device-in-the-iot-hub) created in your Azure IoT Hub using your preferred authentication method* | Have an [individual enrollment](https://docs.microsoft.com/en-us/azure/iot-dps/how-to-manage-enrollments#create-a-device-enrollment) created in your instance of DPS using your preferred authentication method*

>   *While this sample supports SAS keys and Certificates, this guide will refer only to SAS keys.

Reviewer: ```suggestion
  *While this sample supports SAS keys and Certificates, for simplification, this guide will refer only to SAS keys.

```

## Install prerequisites

1. GIT

    Install `git` following the [official website](https://git-scm.com/).

2. ESP-IDF

    Install the [Espressif ESP-IDF](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/get-started/index.html#get-started) following Espressif official documentation (steps 1 to 4).

> 

Author: Done by @wduraes ! Thanks

>     Install the [Espressif ESP-IDF](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/get-started/index.html#get-started) following Espressif official documentation (steps 1 to 4).

Reviewer 2: ```suggestion
    For Windows users, if you don't have [Espressif ESP-IDF](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/get-started/index.html#get-started) installed yet, use [Espressif official installer](https://dl.espressif.com/dl/esp-idf/?idf=4.4), for other Operating Systems or to update an existing installation, follow Espressif official documentation.
```

3. Azure IoT Embedded middleware for FreeRTOS

>     Clone the following repo to download all sample device code, setup scripts, and offline versions of the documentation.

Reviewer: ```suggestion
    Clone the following repo to download all sample device code, setup scripts, and SDK documentation.

```

>     **If you previously cloned this repo in another sample, you don't need to do it again.**

Reviewer: do we realllllly need to call this out??

    ```bash
    git clone https://github.com/Azure-Samples/iot-middleware-freertos-samples.git
    ```

    To initialize the repo, run the following commands:

    ```bash
    cd iot-middleware-freertos-samples
    git submodule update --init --recursive --depth 1
    ```

  You may also need to enable long path support for both Microsoft Windows and git:

>   * Windows: <https://docs.microsoft.com/windows/win32/fileio/maximum-file-path-limitation?tabs=cmd#enable-long-paths-in-windows-10-version-1607-and-later>

Reviewer: ```suggestion
  * Windows: [Enable long paths in Windows 10](https://docs.microsoft.com/windows/win32/fileio/maximum-file-path-limitation?tabs=cmd#enable-long-paths-in-windows-10-version-1607-and-later)
```

>   * Git: as Administrator run `git config --system core.longpaths true`

Reviewer: ```suggestion
  * Git: As administrator, run `git config --system core.longpaths true`

```

>   * ESP-IDF: Espressif also recommends setting the following environment variable on ESP-IDF console as workaround: `IDF_CCACHE_ENABLE=0`

Reviewer 2: ```suggestion
```

## Prepare the sample

> To connect the ESPRESSIF ESP32 to Azure, you will update the sample configuration, build the image, and flash the image to the device.

Reviewer: ```suggestion
To connect the ESPRESSIF ESP32 to Azure, you will update the sample configuration settings, build the image, and flash the image to the device.

```

### Update sample configuration

The configuration of the ESPRESSIF ESP32 sample uses ESP-IDF' samples standard [kconfig](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/kconfig.html) configuration.

On a [console with ESP-IDF](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/get-started/index.html#step-4-set-up-the-environment-variables), run the following commands:

```shell

> cd demos\projects\ESPRESSIF\esp32

Reviewer 3: wrong path

idf.py menuconfig
```

Under menu item `Azure IoT middleware for FreeRTOS Sample Configuration`, update the following configuration values:

Parameter | Value
---------|----------
 `WiFi SSID` | _{Your WiFi SSID}_
 `WiFi Password` | _{Your WiFi password}_

Under menu item `Azure IoT middleware for FreeRTOS Main Task Configuration`, update the following configuration values:

Parameter | Value

> ---------|----------

Author: Configuration Table, add a couple columns explaining what needs to be filled if using DPS or not

Author: This has been replaced with a comment clarifying about the Hub FQDN (unused if using DPS)

 `Use PnP in Azure Sample` | Enabled by default. Disable this option to build a simpler sample without Azure Plug-and-Play.

>  `Azure IoT Hub FQDN` | _{Your Azure IoT Hub Host FQDN}_

Reviewer 3: This is confusing, if you are using DPS, then you dont need to fill out the FQDN or the device ID.

 `Azure IoT Device ID` | _{Your Azure IoT Hub device ID}_
 `Azure IoT Device Symmetric Key` | _{Your Azure IoT Hub device symmetric key}_

>  `Azure IoT Module ID` | _{Your Azure IoT Hub Module ID}_ (IF USING A MODULE; leave blank if not)

Reviewer: ```suggestion
 `Azure IoT Module ID` | _{Your Azure IoT Hub Module ID}_ (optional, specify module id if using a device module; else leave blank if not)

```

> Some parameters contain default values that do not need to be updated.

If you're using **DPS** with an individual enrollment with SAS authentication, set the following parameters:

Parameter | Value
---------|----------
 `Enable Device Provisioning Sample` | _{Check this option to enable DPS in the sample}_
 `Azure Device Provisioning Service ID Scope` | _{Your ID scope value}_
 `Azure Device Provisioning Service Registration ID` | _{Your Device Registration ID value}_

> Some parameters contain default values that do not need to be updated.

Save the configuration (`Shift + S`) inside the sample folder in a file with name `sdkconfig`.
After that, close the configuration utility (`Shift + Q`).

## Build the image

> This step assumes you are in the ESPRESSIF ESP32 sample directory (same as configuration step above).

> 

Reviewer 2: ```suggestion

> On Windows, errors might still occur if a path is too long. Espressif recommends setting the following environment variable on ESP-IDF console `IDF_CCACHE_ENABLE=0` as a workaround.
```
